File: 백준/Gold/5676. 음주 코딩/음주 코딩.py
import sys

def build(node, start, end):
    """ 세그먼트 트리 초기화 """
    if start == end:
        tree[node] = arr[start]
    else:
        mid = (start + end) // 2
        build(2 * node, start, mid)
        build(2 * node + 1, mid + 1, end)
        tree[node] = tree[2 * node] * tree[2 * node + 1]

def update(node, start, end, idx, value):
    """ 값 변경 (C i V) """
    if start == end:
        arr[idx] = value
        tree[node] = value
    else:
        mid = (start + end) // 2
        if idx <= mid:
            update(2 * node, start, mid, idx, value)
        else:
            update(2 * node + 1, mid + 1, end, idx, value)
        tree[node] = tree[2 * node] * tree[2 * node + 1]

def query(node, start, end, l, r):
    """ 곱셈 범위 쿼리 (P i j) """
    if r < start or l > end:
        return 1  # 곱셈에 영향 없는 항등원
    if l <= start and end <= r:
        return tree[node]
    mid = (start + end) // 2
    left_result = query(2 * node, start, mid, l, r)
    right_result = query(2 * node + 1, mid + 1, end, l, r)
    return left_result * right_result

while True:
    try:
        n, k = map(int, sys.stdin.readline().split())
        arr = list(map(int, sys.stdin.readline().split()))
        
        # 1, -1, 0 형태로 변환하여 저장
        arr = [1 if x > 0 else -1 if x < 0 else 0 for x in arr]
        
        tree = [0] * (4 * n)
        build(1, 0, n - 1)

        result = []
        for _ in range(k):
            command = sys.stdin.readline().split()
            if command[0] == "C":
                i, v = int(command[1]) - 1, int(command[2])
                v = 1 if v > 0 else -1 if v < 0 else 0
                update(1, 0, n - 1, i, v)
            else:  # "P i j"
                i, j = int(command[1]) - 1, int(command[2]) - 1
                res = query(1, 0, n - 1, i, j)
                result.append("+" if res > 0 else "-" if res < 0 else "0")

        sys.stdout.write("".join(result) + "\n")

    except:
        break

File: 백준/Gold/5676. 음주 코딩/음주 코딩.java
import java.io.*;
import java.util.*;

public class Main {
    static int[] segTree;
    static int N;
    static int[] arr;
    
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;
        StringBuilder sb = new StringBuilder();
        
        while (true) {
            String line = br.readLine();
            if (line == null || line.isEmpty()) break;
            
            st = new StringTokenizer(line);
            N = Integer.parseInt(st.nextToken());
            int K = Integer.parseInt(st.nextToken());
            
            arr = new int[N];
            segTree = new int[4 * N];
            
            st = new StringTokenizer(br.readLine());
            for (int i = 0; i < N; i++) {
                int x = Integer.parseInt(st.nextToken());
                arr[i] = x > 0 ? 1 : x < 0 ? -1 : 0;
            }
            
            build(1, 0, N - 1);
            
            for (int i = 0; i < K; i++) {
                st = new StringTokenizer(br.readLine());
                String cmd = st.nextToken();
                
                if (cmd.equals("C")) {
                    int index = Integer.parseInt(st.nextToken()) - 1;
                    int value = Integer.parseInt(st.nextToken());
                    value = value > 0 ? 1 : value < 0 ? -1 : 0;
                    update(1, 0, N - 1, index, value);
                } else {
                    int l = Integer.parseInt(st.nextToken()) - 1;
                    int r = Integer.parseInt(st.nextToken()) - 1;
                    int result = query(1, 0, N - 1, l, r);
                    sb.append(result > 0 ? "+" : result < 0 ? "-" : "0");
                }
            }
            sb.append("\n");
        }
        System.out.print(sb);
    }
    
    static void build(int node, int start, int end) {
        if (start == end) {
            segTree[node] = arr[start];
        } else {
            int mid = (start + end) / 2;
            build(2 * node, start, mid);
            build(2 * node + 1, mid + 1, end);
            segTree[node] = segTree[2 * node] * segTree[2 * node + 1];
        }
    }
    
    static void update(int node, int start, int end, int idx, int value) {
        if (start == end) {
            arr[idx] = value;
            segTree[node] = value;
        } else {
            int mid = (start + end) / 2;
            if (idx <= mid) {
                update(2 * node, start, mid, idx, value);
            } else {
                update(2 * node + 1, mid + 1, end, idx, value);
            }
            segTree[node] = segTree[2 * node] * segTree[2 * node + 1];
        }
    }
    
    static int query(int node, int start, int end, int l, int r) {
        if (r < start || l > end) return 1;
        if (l <= start && end <= r) return segTree[node];
        int mid = (start + end) / 2;
        return query(2 * node, start, mid, l, r) * query(2 * node + 1, mid + 1, end, l, r);
    }
}